File: blf_lib/src/blam/haloreach/v12065_11_08_24_1738_tu1actual/game/game_engine_default.rs
use serde::{Deserialize, Serialize};
use blf_lib::blam::haloreach::v12065_11_08_24_1738_tu1actual::game::game_engine_player_traits::c_player_traits;
use blf_lib::io::bitstream::{c_bitstream_reader, c_bitstream_writer};
use blf_lib::{SET_BIT, TEST_BIT};
use blf_lib::blam::haloreach::v12065_11_08_24_1738_tu1actual::game::game_engine_loadout_traits::c_game_engine_loadout_traits;
use blf_lib::blam::haloreach::v12065_11_08_24_1738_tu1actual::game::game_engine_team::c_game_engine_team_options;
use blf_lib::blam::haloreach::v12065_11_08_24_1738_tu1actual::game::game_engine_traits::{c_game_engine_miscellaneous_options, c_game_engine_respawn_options};
use blf_lib::blam::haloreach::v12065_11_08_24_1738_tu1actual::saved_games::saved_game_files::s_content_item_metadata;
use blf_lib_derivable::result::BLFLibResult;

pub const k_game_engine_type_count: usize = 11;

#[derive(Default, PartialEq, Debug, Clone, Serialize, Deserialize)]
pub struct c_game_engine_social_options {
    pub m_flags: u16,
    pub m_team_changing: u16,
    pub m_observers_enabled: bool,
}

impl c_game_engine_social_options {
    pub fn encode(&self, bitstream: &mut c_bitstream_writer) -> BLFLibResult {
        bitstream.write_bool(self.m_observers_enabled)?; // TODO: what is this
        bitstream.write_integer(self.m_team_changing as u32, 2)?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 0))?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 1))?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 2))?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 3))?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 4))?;

        Ok(())
    }

    pub fn decode(&mut self, bitstream: &mut c_bitstream_reader) -> BLFLibResult {
        self.m_observers_enabled = bitstream.read_unnamed_bool()?;
        self.m_team_changing = bitstream.read_unnamed_integer(2)?;
        SET_BIT!(self.m_flags, 0, bitstream.read_unnamed_bool()?);
        SET_BIT!(self.m_flags, 1, bitstream.read_unnamed_bool()?);
        SET_BIT!(self.m_flags, 2, bitstream.read_unnamed_bool()?);
        SET_BIT!(self.m_flags, 3, bitstream.read_unnamed_bool()?);
        SET_BIT!(self.m_flags, 4, bitstream.read_unnamed_bool()?);

        Ok(())
    }
}

#[derive(Default, PartialEq, Debug, Clone, Serialize, Deserialize)]
pub struct c_game_engine_map_override_options {
    pub m_flags: u32,
    pub m_base_player_traits: c_player_traits,
    pub m_weapon_set_absolute_index: i16,
    pub m_vehicle_set_absolute_index: i16,
    pub m_red_powerup_traits: c_player_traits,
    pub m_blue_powerup_traits: c_player_traits,
    pub m_yellow_powerup_traits: c_player_traits,
    pub m_red_powerup_duration_seconds: u8,
    pub m_blue_powerup_duration_seconds: u8,
    pub m_yellow_powerup_duration_seconds: u8,
}

impl c_game_engine_map_override_options {
    pub fn encode(&self, bitstream: &mut c_bitstream_writer) -> BLFLibResult {
        bitstream.write_bool(TEST_BIT!(self.m_flags, 0))?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 1))?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 2))?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 3))?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 4))?;
        bitstream.write_bool(TEST_BIT!(self.m_flags, 5))?;
        self.m_base_player_traits.encode(bitstream)?;
        bitstream.write_signed_integer(self.m_weapon_set_absolute_index as i32, 8)?;
        bitstream.write_signed_integer(self.m_vehicle_set_absolute_index as i32, 8)?;
        self.m_red_powerup_traits.encode(bitstream)?;
        self.m_blue_powerup_traits.encode(bitstream)?;
        self.m_yellow_powerup_traits.encode(bitstream)?;
        bitstream.write_integer(self.m_red_powerup_duration_seconds as u32, 7)?;
        bitstream.write_integer(self.m_blue_powerup_duration_seconds as u32, 7)?;
        bitstream.write_integer(self.m_yellow_powerup_duration_seconds as u32, 7)?;

        Ok(())
    }

    pub fn decode(&mut self, bitstream: &mut c_bitstream_reader) -> BLFLibResult {
        self.m_flags = bitstream.read_integer("flags", 6)?;
        self.m_base_player_traits.decode(bitstream)?;
        self.m_weapon_set_absolute_index = bitstream.read_signed_integer("map-override-weapon-set", 8)?;
        self.m_vehicle_set_absolute_index = bitstream.read_signed_integer("map-override-vehicle-set", 8)?;
        self.m_red_powerup_traits.decode(bitstream)?;
        self.m_blue_powerup_traits.decode(bitstream)?;
        self.m_yellow_powerup_traits.decode(bitstream)?;
        self.m_red_powerup_duration_seconds = bitstream.read_integer("map-override-red-powerup-duration", 7)?;
        self.m_blue_powerup_duration_seconds = bitstream.read_integer("map-override-blue-powerup-duration", 7)?;
        self.m_yellow_powerup_duration_seconds = bitstream.read_integer("map-override-yellow-powerup-duration", 7)?;

        Ok(())
    }
}


#[derive(Default, PartialEq, Debug, Clone, Serialize, Deserialize)]
pub struct c_game_engine_base_variant {
    pub m_metadata: s_content_item_metadata,
    pub m_built_in: bool,
    pub m_miscellaneous_options: c_game_engine_miscellaneous_options,
    pub m_respawn_options: c_game_engine_respawn_options,
    pub m_social_options: c_game_engine_social_options,
    pub m_map_override_options: c_game_engine_map_override_options,
    pub m_team_scoring_method: u16,
    pub m_team_options: c_game_engine_team_options,
    pub m_loadouts: c_game_engine_loadout_traits,

}

impl c_game_engine_base_variant {
    pub fn encode(&self, bitstream: &mut c_bitstream_writer) -> BLFLibResult {
        self.m_metadata.encode(bitstream)?;
        bitstream.write_bool(self.m_built_in)?;
        self.m_miscellaneous_options.encode(bitstream)?;
        self.m_respawn_options.encode(bitstream)?;
        self.m_social_options.encode(bitstream)?;
        self.m_map_override_options.encode(bitstream)?;
        bitstream.write_integer(self.m_team_scoring_method, 3)?;
        self.m_team_options.encode(bitstream)?;
        self.m_loadouts.encode(bitstream)?;

        Ok(())
    }

    pub fn decode(&mut self, bitstream: &mut c_bitstream_reader) -> BLFLibResult {
        self.m_metadata.decode(bitstream)?;
        self.m_built_in = bitstream.read_bool("variant-built-in")?;
        self.m_miscellaneous_options.decode(bitstream)?;
        self.m_respawn_options.decode(bitstream)?;
        self.m_social_options.decode(bitstream)?;
        self.m_map_override_options.decode(bitstream)?;
        self.m_team_scoring_method = bitstream.read_integer("team-scoring-method", 3)?;
        self.m_team_options.decode(bitstream)?;
        self.m_loadouts.decode(bitstream)?;


        Ok(())
    }
}

File: blf-ts/src/blam/halo3odst_mcc/v_untracked_25_08_16_1402/game/game_engine_default.ts
import type {
  c_bitstream_reader,
  c_bitstream_writer,
} from "../../../../bitstream";
import { AutoMap } from "../../../../helpers/automap";
import { s_content_item_metadata } from "../saved_games/saved_game_files";
import { c_player_traits } from "./game_engine_player_traits";
import {
  c_game_engine_miscellaneous_options,
  c_game_engine_respawn_options,
} from "./game_engine_traits";

function testBit(flags: number, bit: number): boolean {
  return (flags & (1 << bit)) !== 0;
}

function setBit(flags: number, bit: number, value: boolean): number {
  if (value) {
    return flags | (1 << bit);
  }
  return flags & ~(1 << bit);
}

export const k_game_engine_type_count = 12;

/** Port of Halo 3 12070 `c_game_engine_social_options`. */
export class c_game_engine_social_options {
  @AutoMap(() => Number)
  m_flags = 0;
  @AutoMap(() => Number)
  m_team_changing = 0;

  encode(bitstream: c_bitstream_writer): void {
    bitstream.write_bool(false);
    bitstream.write_integer(this.m_team_changing, 2);
    bitstream.write_bool(testBit(this.m_flags, 0));
    bitstream.write_bool(testBit(this.m_flags, 1));
    bitstream.write_bool(testBit(this.m_flags, 2));
    bitstream.write_bool(testBit(this.m_flags, 3));
    bitstream.write_bool(testBit(this.m_flags, 4));
  }

  decode(bitstream: c_bitstream_reader): void {
    bitstream.seek_relative(1);
    this.m_team_changing = bitstream.read_integer("team-changing", 2);
    this.m_flags = setBit(
      this.m_flags,
      0,
      bitstream.read_bool("social-options-flag-0")
    );
    this.m_flags = setBit(
      this.m_flags,
      1,
      bitstream.read_bool("social-options-flag-1")
    );
    this.m_flags = setBit(
      this.m_flags,
      2,
      bitstream.read_bool("social-options-flag-2")
    );
    this.m_flags = setBit(
      this.m_flags,
      3,
      bitstream.read_bool("social-options-flag-3")
    );
    this.m_flags = setBit(
      this.m_flags,
      4,
      bitstream.read_bool("social-options-flag-4")
    );
  }
}

/** Port of Halo 3 12070 `c_game_engine_map_override_options`. */
export class c_game_engine_map_override_options {
  @AutoMap(() => Number)
  m_flags = 0;
  @AutoMap(() => c_player_traits)
  m_base_player_traits = new c_player_traits();
  @AutoMap(() => Number)
  m_weapon_set_absolute_index = 0;
  @AutoMap(() => Number)
  m_vehicle_set_absolute_index = 0;
  @AutoMap(() => c_player_traits)
  m_red_powerup_traits = new c_player_traits();
  @AutoMap(() => c_player_traits)
  m_blue_powerup_traits = new c_player_traits();
  @AutoMap(() => c_player_traits)
  m_yellow_powerup_traits = new c_player_traits();
  @AutoMap(() => Number)
  m_red_powerup_duration_seconds = 0;
  @AutoMap(() => Number)
  m_blue_powerup_duration_seconds = 0;
  @AutoMap(() => Number)
  m_yellow_powerup_duration_seconds = 0;

  encode(bitstream: c_bitstream_writer): void {
    bitstream.write_bool(testBit(this.m_flags, 0));
    bitstream.write_bool(testBit(this.m_flags, 1));
    this.m_base_player_traits.encode(bitstream);
    bitstream.write_signed_integer(this.m_weapon_set_absolute_index, 8);
    bitstream.write_signed_integer(this.m_vehicle_set_absolute_index, 8);
    this.m_red_powerup_traits.encode(bitstream);
    this.m_blue_powerup_traits.encode(bitstream);
    this.m_yellow_powerup_traits.encode(bitstream);
    bitstream.write_integer(this.m_red_powerup_duration_seconds, 7);
    bitstream.write_integer(this.m_blue_powerup_duration_seconds, 7);
    bitstream.write_integer(this.m_yellow_powerup_duration_seconds, 7);
  }

  decode(bitstream: c_bitstream_reader): void {
    this.m_flags = setBit(
      this.m_flags,
      0,
      bitstream.read_bool("map-override-flag-0")
    );
    this.m_flags = setBit(
      this.m_flags,
      1,
      bitstream.read_bool("map-override-flag-1")
    );
    this.m_base_player_traits.decode(bitstream);
    this.m_weapon_set_absolute_index = bitstream.read_signed_integer(
      "weapon-set-absolute-index",
      8
    );
    this.m_vehicle_set_absolute_index = bitstream.read_signed_integer(
      "vehicle-set-absolute-index",
      8
    );
    this.m_red_powerup_traits.decode(bitstream);
    this.m_blue_powerup_traits.decode(bitstream);
    this.m_yellow_powerup_traits.decode(bitstream);
    this.m_red_powerup_duration_seconds = bitstream.read_integer(
      "red-powerup-duration-seconds",
      7
    );
    this.m_blue_powerup_duration_seconds = bitstream.read_integer(
      "blue-powerup-duration-seconds",
      7
    );
    this.m_yellow_powerup_duration_seconds = bitstream.read_integer(
      "yellow-powerup-duration-seconds",
      7
    );
  }
}

export class c_game_engine_base_variant {
  @AutoMap(() => Number)
  m_checksum = 0;
  @AutoMap(() => s_content_item_metadata)
  m_metadata = new s_content_item_metadata();
  @AutoMap(() => c_game_engine_miscellaneous_options)
  m_miscellaneous_options = new c_game_engine_miscellaneous_options();
  @AutoMap(() => c_game_engine_respawn_options)
  m_respawn_options = new c_game_engine_respawn_options();
  @AutoMap(() => c_game_engine_social_options)
  m_social_options = new c_game_engine_social_options();
  @AutoMap(() => c_game_engine_map_override_options)
  m_map_override_options = new c_game_engine_map_override_options();
  @AutoMap(() => Number)
  m_flags = 0;
  @AutoMap(() => Number)
  m_team_scoring_method = 0;

  encode(bitstream: c_bitstream_writer): void {
    this.m_metadata.encode(bitstream);
    bitstream.write_integer(this.m_flags, 1);
    this.m_miscellaneous_options.encode(bitstream);
    this.m_respawn_options.encode(bitstream);
    this.m_social_options.encode(bitstream);
    this.m_map_override_options.encode(bitstream);
    bitstream.write_integer(this.m_team_scoring_method, 3);
  }

  decode(bitstream: c_bitstream_reader): void {
    this.m_metadata.decode(bitstream);
    this.m_flags = bitstream.read_integer("base-variant-flags", 1);
    this.m_miscellaneous_options.decode(bitstream);
    this.m_respawn_options.decode(bitstream);
    this.m_social_options.decode(bitstream);
    this.m_map_override_options.decode(bitstream);
    this.m_team_scoring_method = bitstream.read_integer(
      "team-scoring-method",
      3
    );
  }
}
